### ISSUE TITLE
❓ [QUESTION] What exactly do we get from expression our input in a spherical basis?

### ISSUE THREAD

**Issue by tueboesen** [e3nn/e3nn#269]
I have been trying to convince a professor I'm working with about the value of this work, and why I believe equivariance to be important and why we should use the spherical harmonics. After weeks of discussion, I think I have finally convinced him of the importance of equivariance in networks in some cases, but he is so far very against expressing vectors in spherical harmonics and claims that all of that isn't important and he can do all the needed operations equivariantly directly on the vectors, as done in [this paper](https://arxiv.org/pdf/2102.09844.pdf), where they build the equivariant neural network from simple weighted vector additions. (One of the main problems is that he is a professor in mathematics with no physics background, so to him a tensorproduct is an outer product of two vectors and anything else is highly suspect).

Now what remains unclear to me at this point is whether expressing the input parameters in spherical harmonics and performing the tensor products, actually leads to more sophisticated equivariant operations (I believe it does, but I'm not entirely sure, and I can't show it/prove it). Another point might be about the faithful physical representation of various input/output parameters, where your work allows the specification of whether parameters are vectors, pseudo-vectors, scalars or pseudo-scalars. However this is not limited to spherical harmonics either, though the interactions of them is certainly made easier in spherical harmonics.

So all in all, what are the main benefits of expressing variables in terms of spherical harmonics?, what does this make possible, that would otherwise be impossible/very hard to do?

**Comment by blondegeek** [comment]
Hi Tue,

You certainly could accomplish the same expressivity by performing outer
products in for example a Cartesian basis -- the issue is that you then
need more and more indices. The irrep basis (the spherical harmonic basis +
the irreps with opposite parity than the spherical harmonics) is a provably
complete basis -- ANYTHING that exists in Euclidean 3D space can be
expressed as a direct sum of irreps. This means you can express a Cartesian
tensor of any rank (with any number of indices) as a single index irrep
tensor. Having a single index is handy because we use composition in neural
networks -- our outputs from one layer become the inputs to another layer.

With regard to the specific paper you referenced, E(n) equivariant graph
neural networks, from at least my initial reading, my understanding is the
following:
1) It's highest order representation is a vector representation

...with which you could certain perform outer products to create higher
rank objects as output but the internal representation is only vectors

2) it only uses scalar operations.

For example, operations on vectors are scalar in nature -- of the form
vector x scalar -> vector. There are no cross-products (vector x vector ->
vector) or higher order operations (vector x vector -> L=2) in the linear
parts of the network. A possible way to test whether I'm correct in this
line of thinking is to set up a toy example where there are charged
particles in a plane (e.g. the xy plane) with initial velocities in the
same plane and then predict forces that are perpendicular to that plane. If
I am correct that the network does not contain cross-products -- then the
networks will not be able to do the task.


If I am in error on any of these points -- I would very much like to know!
So please feel free to help me by correcting my understanding of the paper.

So in summary, the expressivity of e3nn comes from the fact that we use
generalized tensor products (not just scalar operations) which operate on
equivariant objects of arbitrary "rank" (in the Cartesian tensor sense of
having many indices) or "order" (in the irrep sense of L's). Whether this
level of generality is necessary for a specific training task is situation
specific.

Is that helpful? Like I said above, please let me know if there are holes
in my understanding of the paper you referenced and/or if you have any
further questions.

Best!
Tess

**Comment by mariogeiger** [comment]
Dear Tue,

What do you mean precisely when you say
> expressing vectors in spherical harmonics

**Comment by tueboesen** [comment]
Thank you Tess for the long and detailed reply!

I believe your understanding of the E(n) equivariant graph neural networks paper is correct, and I think that the network would indeed fail to predict a cross product out of a 2D plane, if the input solely spans the 2D plane. 

I showed your comment to my collaborating professor, and while I think he ended up agreeing that the E(n) network would not be able to predict the cross-product, he said that this was an edge case which he believes is not going to affect the actual performance in most real world applications, and unless you specifically need the higher order components either for input or output he believes that the higher order interactions would be irrelevant for the predictive power of the network. In our case the current application is protein folding, and for that task we effectively only have scalar input and wish to predict coordinates or inter-atomic distances as output, so in a certain sense he is clearly right that everything we need should be possible with the E(n) network, the question is of course whether the higher express-ability of E3nn actually lead to improved predictions over E(n) in this particular use case, which I believe it does, but I guess we will have to see.

So yeah your comment was very helpful, it unfortunately didn't sway my collaborator, but at least it progressed our discussion, and I think we will just have to try both networks and see how they perform in comparison.

**Comment by tueboesen** [comment]
> What do you mean precisely when you say
> 
> > expressing vectors in spherical harmonics
> 

I'm sorry I wasn't very precise. 
When I wrote that I had this specific piece of code in mind:
`edge_sh = o3.spherical_harmonics(self.irreps_edge_attr, edge_vec, True, normalization='component')`
Which if my understanding is correct, is a function that takes a vector as input and outputs the vectors representation in the 2l+1 dimensional space defined by the real spherical harmonics up to order l. I am not sure whether to view this as a change of basis or merely an embedding in a higher dimensional space? 

Looking more closely at that code I also notice that it only takes vectors as input. My understanding of the e3nn framework (and based on Tess' comment above) is that it should also allow Cartesian rank 2 (or higher) tensors as input right? As in, I should be able to use for instance a stress tensor as input? Though I guess that would need a different function. 

> PS: where did you read about that about our work?

I stumbled upon Tess'  [talk](https://www.youtube.com/watch?v=VN2biLjqJXc ) about a month ago, 
which lead me to [this paper](https://arxiv.org/pdf/2101.03164.pdf), and from there I found the [previous work](https://arxiv.org/pdf/1802.08219.pdf).
I clearly don't have that firm a grasp on e3nn or the underlying theory yet, but I replicated the network and performance (well almost, but I didn't want to train for 8 days :) ) on aspirin from the MD17 molecules, which helped me learn a lot about how e3nn works.

**Comment by mariogeiger** [comment]
> I am not sure whether to view this as a change of basis or merely an embedding in a higher dimensional space?

these are polynomials of increasing order `0, 1, 2..., l`

>  it should also allow Cartesian rank 2 (or higher) tensors as input right? 

Indeed, basically any representation can be converted (via a change of basis) into the irreps and then be fed into an e3nn network.

> As in, I should be able to use for instance a stress tensor as input?

Yes, if your stress tensor is symmetric you have one scalar and one even L=2 irrep in it and the code below gives you the change of basis.
```
r = o3.ReducedTensorProducts('ij=ji', i='1o')
r.irreps_out  # prints 1x0e+1x2e
r.change_of_basis  # [1+5, 3, 3] tensor
```

> I stumbled upon Tess' talk about a month ago,

Cool, good to know from where the users come from :-)
